### PR TITLE
fix(internal) cdef Python exceptions being ignored

### PR DESCRIPTION
**Pull request by brettlangdon**
## Commit Message
{{title}}

In a cdef function `raise` statements will create a Python exception, add it to the exception stack, and then immediately return. However, the caller will not explicitly check to see if any exception has occurred unless an exception return value has been provided.

Until now these cdef functions when they encountered a `raise` would push the Python exception, and then return the default `int` value of `0`. Which based on how we use the return values `0` means success.

This means that anytime we encounter an exception in our encoder we potentially ignore any exceptions and still append the partially encoded trace into the buffer.

With this change we are standardizing around `except? -1` which tells the caller that their _might_ be a Python exception if the return value is `-1`. We do this because the `msgpack_*` functions we call will also return `-1` on error, but will not create a Python exception.

### CODE REVIEW

**Comment by brettlangdon** [comment]
This is basically the same as #2966 but using `except? -1` and a slightly different testing strategy.

The test case added isn't as robust as it probably needs to be.

**Review by P403n1x87 — DISMISSED** [review]

**Review by mabdinur — APPROVED** [review]

**Comment by brettlangdon** [comment]
@Mergifyio backport 0.56

File: releasenotes/notes/fix-encoding-exceptions-c12900b38741d2bf.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+fixes:
+  - |
+    Fix Python exceptions during Trace encoding being ignored and creating invalid payloads.
```
Comment (majorgreys) [inline releasenotes/notes/fix-encoding-exceptions-c12900b38741d2bf.yaml:4]:
```suggestion
    Fix handling of Python exceptions during trace encoding. The tracer will no longer silently fail to encode invalid span data and instead log an exception.
```

**Review by majorgreys — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]
As discussed offline we should reevaluate the error handling that we do here in general as a follow up.

For example, I noticed in `get_bytes` that it's a `cdef` that can possibly return `None` on error. We do handle the `None` return in `AgentWriter()` but we don't log or raise anything which would allow errors to silently occur 😕

**Pull request merged**